gapis/trace/android/trace.go

```diff
@@ -215,32 +224,62 @@ func (t *androidTracer) Validate(ctx context.Context) error {
 
 		startFunc(ctx)
 		if !doneSignal.Wait(ctx) {
-			err = log.Err(ctx, err, "Fail to wait for done signal from Perfetto.")
+			err = log.Err(ctx, err, "Failed to wait for done signal from Perfetto.")
 		}
 		log.I(ctx, "Perfetto trace size %v bytes", written)
 	})
 	if err != nil {
-		return err
+		return nil, err
+	}
+
+	// Attempt to download perfetto trace.
+	var temp file.Path
+	if enableLocalFiles {
+		temp, err = file.TempWithExt("validation", "perfetto")
+		if err != nil {
+			return nil, log.Err(ctx, err, "Failed to create a temp file for trace")
+		}
 	}
 
 	var processor *perfetto.Processor
+	var traceLoadingErr error
 	status.Do(ctx, "Trace loading", func(ctx context.Context) {
-		// Load Perfetto trace and create trace processor.
-		if config.DumpValidationTrace {
-			if out, err := os.Create("validate.perfetto"); err == nil {
-				out.Write(buf.Bytes())
-				out.Close()
-			}
-		}
 		processor, err = perfetto.NewProcessor(ctx, buf.Bytes())
+		if err != nil {
+			traceLoadingErr = log.Err(ctx, err, "Failed to initialize the perfetto processor")
+			return
+		}
+
+		file, err := os.OpenFile(temp.System(), os.O_APPEND|os.O_WRONLY, fs.ModeAppend)
+		if err != nil {
+			traceLoadingErr = log.Err(ctx, err, "Failed to open temp file")
+			return
+		}
+		defer file.Close()
+
+		numWritten, err := file.Write(buf.Bytes())
+		if err != nil {
+			traceLoadingErr = log.Err(ctx, err, "Failed to write trace to temp file")
+			return
+		}
+		log.I(ctx, "Writing trace size %v bytes to %v", numWritten, file.Name())
 	})
-	if err != nil {
-		return err
+	res := &service.DeviceValidationResult{
+		TracePath: temp.System(),
 	}
-	defer processor.Close()
 	ctx = status.Start(ctx, "Validation")
+	if traceLoadingErr != nil {
+		return nil, traceLoadingErr
+	}
+	defer processor.Close()
```

Comment:
These two defers are now in the wrong place. the `defer processor.Close()` should be right after the first error check after the processor is created, line 252. `defer status.Finish(ct)` should be right after the `status.Start` line, otherwise we'd not call finish if `traceLodingErr` is not nil. Maybe move that check to before the `status` stuff?